File: backend/internal/finance/app/persons.go
package app

import (
	"context"
	"strings"

	"github.com/google/uuid"

	"github.com/corsi/backend/internal/finance/domain"
	"github.com/corsi/backend/internal/finance/ports"
)

type CreatePersonInput struct {
	WorkspaceID uuid.UUID
	Name        string
	Notes       string
}

func (s *Service) CreatePerson(ctx context.Context, in CreatePersonInput) (*domain.Person, error) {
	p := &domain.Person{
		ID:          uuid.New(),
		WorkspaceID: in.WorkspaceID,
		Name:        strings.TrimSpace(in.Name),
		Notes:       in.Notes,
	}
	if err := p.Validate(); err != nil {
		return nil, err
	}
	if err := s.repos.Persons.Create(ctx, p); err != nil {
		return nil, err
	}
	return p, nil
}

type UpdatePersonInput struct {
	WorkspaceID uuid.UUID
	ID          uuid.UUID
	Name        *string
	Notes       *string
}

func (s *Service) UpdatePerson(ctx context.Context, in UpdatePersonInput) (*domain.Person, error) {
	current, err := s.repos.Persons.FindByID(ctx, in.WorkspaceID, in.ID)
	if err != nil {
		return nil, err
	}
	if in.Name != nil {
		current.Name = strings.TrimSpace(*in.Name)
	}
	if in.Notes != nil {
		current.Notes = *in.Notes
	}
	if err := current.Validate(); err != nil {
		return nil, err
	}
	if err := s.repos.Persons.Update(ctx, current); err != nil {
		return nil, err
	}
	return current, nil
}

func (s *Service) DeletePerson(ctx context.Context, workspaceID, id uuid.UUID) error {
	return s.repos.Persons.SoftDelete(ctx, workspaceID, id)
}

func (s *Service) GetPerson(ctx context.Context, workspaceID, id uuid.UUID) (*domain.Person, error) {
	return s.repos.Persons.FindByID(ctx, workspaceID, id)
}

type ListPersonsInput struct {
	WorkspaceID uuid.UUID
	Limit       int
	Offset      int
}

func (s *Service) ListPersons(ctx context.Context, in ListPersonsInput) ([]domain.Person, error) {
	return s.repos.Persons.List(ctx, in.WorkspaceID, ports.PersonFilter{Limit: in.Limit, Offset: in.Offset})
}

File: frontend/src/modules/finance/api/persons.ts
/**
 * Typed REST client for /finance/persons.
 *
 * Backend models name + notes only. Frontend extras (role, active,
 * whatsappNumber) ride a local sidecar — see `personMetadata.ts`.
 *
 * Wire shape matches `backend/api/openapi/finance.yaml` and the Go
 * struct in `backend/internal/finance/domain/person.go`.
 */

import { apiFetch, type PageEnvelope } from "@/lib/api/client";

export interface ApiPerson {
  id: string;
  workspace_id: string;
  name: string;
  notes: string;
  created_at: string;
  updated_at: string;
  deleted_at?: string | null;
}

export interface CreatePersonRequest {
  name: string;
  notes?: string;
}

export interface UpdatePersonRequest {
  name?: string;
  notes?: string;
}

export interface ListPersonsParams {
  limit?: number;
  offset?: number;
}

function qs(params: object): string {
  const sp = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined) sp.set(k, String(v));
  }
  const s = sp.toString();
  return s ? `?${s}` : "";
}

export async function listPersons(p: ListPersonsParams = {}): Promise<ApiPerson[]> {
  const env = await apiFetch<PageEnvelope<ApiPerson>>(
    `/finance/persons${qs({ limit: 100, ...p })}`,
  );
  return env.items;
}

export function getPerson(id: string): Promise<ApiPerson> {
  return apiFetch<ApiPerson>(`/finance/persons/${id}`);
}

export function createPerson(body: CreatePersonRequest): Promise<ApiPerson> {
  return apiFetch<ApiPerson>(`/finance/persons`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function updatePerson(id: string, body: UpdatePersonRequest): Promise<ApiPerson> {
  return apiFetch<ApiPerson>(`/finance/persons/${id}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export function deletePerson(id: string): Promise<void> {
  return apiFetch<void>(`/finance/persons/${id}`, { method: "DELETE" });
}
